### ISSUE TITLE
Mac tests are flaking due to missing `runtime`

### ISSUE THREAD

**Issue by keyonghan** [flutter/flutter#99429]
The bot [build807-m9](https://chromium-swarm.appspot.com/bot?id=build807-m9) starts causing test flake after a release build failure: https://luci-milo.appspot.com/ui/p/flutter/builders/prod/Mac%20beta%20verify_binaries_codesigned/400/overview

![Screen Shot 2022-03-02 at 12 44 14 PM](https://user-images.githubusercontent.com/54558023/156446664-9e7d8622-c0cb-4c2d-9c7b-439a80bf3313.png)

All these failures are missing `Runtimes` package, e.g. https://ci.chromium.org/ui/p/flutter/builders/prod/Mac%20module_test_ios/5288/overview

It is possible the cache need a clean up.

**Comment by keyonghan** [comment]
Marking as P1 as it causes multiple flakes across different tests.

**Comment by godofredoc** [comment]
aaaah, if those tests are not using the default runtimes it is failing because we don't have support from that yet. We'll need to map a .ci.yaml property to the runtimes property in the xcode module.

**Comment by keyonghan** [comment]
> aaaah, if those tests are not using the default runtimes it is failing because we don't have support from that yet. We'll need to map a .ci.yaml property to the runtimes property in the xcode module.

I don't think we are using customized runtime for existing tests. All should be using the default ones? /cc @jmagman 

The issue is happening on build811 as well. Almost all flakes on the recent runs are due to these two bots
![Screen Shot 2022-03-03 at 9 59 24 AM](https://user-images.githubusercontent.com/54558023/156634071-90385887-c878-4ef1-b00d-323b4e80c372.png)

/cc @CaseyHillers can we move these two bots to staging during investigation?

**Comment by jmagman** [comment]
> > aaaah, if those tests are not using the default runtimes it is failing because we don't have support from that yet. We'll need to map a .ci.yaml property to the runtimes property in the xcode module.
> 
> I don't think we are using customized runtime for existing tests. All should be using the default ones? /cc @jmagman

Correct, the framework repo is only using the default iOS 15 runtime.  Which should be handled by the mac_toolchain populating the Xcode cache.  This looks the same as https://github.com/flutter/flutter/issues/99039 ?

**Comment by keyonghan** [comment]
Same issue is happening on build810 too: https://chromium-swarm.appspot.com/bot?id=build810-m9

All these three bots started hitting runtine package issue after a failure on a releaser builder, e.g.
![Screen Shot 2022-03-03 at 4 32 30 PM](https://user-images.githubusercontent.com/54558023/156676511-d30af68e-5719-416a-a29a-4b70f044314b.png)

**Comment by jmagman** [comment]
When you download a vanilla version of Xcode, there's a default iOS simulator located in `Contents/Developer/Platforms/iPhoneOS.platform/Library/Developer/CoreSimulator/Profiles/Runtimes` (iOS 15 in this case).  However when `mac_toolchain` uploads the Xcode bundle, it strips that runtime out.  The runtime is uploaded as a separate cipd package `ios_runtime`.  See https://source.chromium.org/chromium/infra/infra/+/main:go/src/infra/cmd/mac_toolchain/README.md and https://g3doc.corp.google.com/company/teams/chrome/ios/docs/infra/xcode_upload.md?cl=head#upload-ios-runtimes

So how this should work is when the Xcode cache is first by
```
  '/opt/s/w/ir/cache/osx_sdk/mac_toolchain',
  'install',
  '-kind',
  'ios',
  '-xcode-version',
  '12d4e',
  '-output-dir',
  '/opt/s/w/ir/cache/osx_sdk/XCode.app',
```
it should download three packages, `flutter_internal/ios/xcode/mac`, `flutter_internal/ios/xcode/ios` (one of those the Xcode bundle with nothing in that directory), and `flutter_internal/ios/xcode/ios_runtime` (which just has the runtime, and puts it in the correct `Contents/Developer/Platforms/iPhoneOS.platform/Library/Developer/CoreSimulator/Profiles/Runtimes` directory.
```
[P92769 18:56:42.815 action_plan.go:160 I] In root:
[P92769 18:56:42.815 action_plan.go:166 I]   to install:
[P92769 18:56:42.815 action_plan.go:168 I]     flutter_internal/ios/xcode/mac:UCaNcftiBhJwpRpc5DUfKSIw8COAV-Tys6P58EzPEykC
[P92769 18:56:42.815 action_plan.go:168 I]     flutter_internal/ios/xcode/ios:rWdxdSMwqPafHwtN9Uh77do6HVmD6iVatO_XNPbfHvEC
...
[P92815 19:01:13.007 action_plan.go:160 I] In root:
[P92815 19:01:13.007 action_plan.go:166 I]   to install:
[P92815 19:01:13.007 action_plan.go:168 I]     flutter_internal/ios/xcode/ios_runtime:xbExNWHTJfpNf8vDNXhmLNx9sjfFvAl4ed74gTrXl3cC
```
(copied from a `led` run but it should look like this everywhere when the Xcode cache is populated). https://logs.chromium.org/logs/flutter/led/magder_google.com/34b0add45b8fe97daadbfb8f634b2426c6fdea1e39c03af96488ca5ff59eb6f2/+/u/install_xcode/stdout

So either the `ios_runtime` part isn't being downloaded, or it is but it's somehow being stripped out of the cache, or maybe there's something wrong with `ios-15` runtime package as @keyonghan suggested in https://github.com/flutter/flutter/issues/99039#issuecomment-1051109880 (pretty sure these are the same issues).  However if there was something wrong with the runtime I would expect these tests to fail on every builder, not just some.

Some questions to answer:
1. On a machine where this is happening, is `Contents/Developer/Platforms/iPhoneOS.platform/Library/Developer/CoreSimulator/Profiles/Runtimes` empty?
2. In the original build where the Xcode cache was populated, does it show that `ios_runtime` download in the `install_xcode` step?
3. Is the `mac_toolchain` version correct on these machines? The runtime feature was introduced not that long ago. https://flutter-review.googlesource.com/c/recipes/+/23980
4. The framework does not need custom runtimes, it can use the default iOS 15, but is it possible the changes to support that in the engine like https://flutter-review.googlesource.com/c/recipes/+/24944 separated out the caches, and the framework needs to add the `xcode_runtime_ios-15-0` cache explicitly somewhere? @godofredoc may have insights here.

**Comment by godofredoc** [comment]
If a build using custom runtimes ran on the failing test bed it deletes all the runtimes and installs the one passed explicitly. Because the next executions do not pass a custom runtime and the original xcode was already cached it does not install the cipd packages again and only the custom runtime installed previously is available.

https://cs.opensource.google/flutter/recipes/+/main:recipe_modules/osx_sdk/api.py;l=214

**Comment by jmagman** [comment]
> If a build using custom runtimes ran on the failing test bed it deletes all the runtimes and installs the one passed explicitly.

This is happening in prod, and there aren't any custom runtimes in prod, only the "default" one.

> only the custom runtime installed previously is available.

Though it looks like no runtimes are available, though still want to know what's in `Contents/Developer/Platforms/iPhoneOS.platform/Library/Developer/CoreSimulator/Profiles/Runtimes` to know if the runtime is there and bad somehow, or if the directory is empty.

**Comment by CaseyHillers** [comment]
I ran a LED test specifying the runtime for a failed mac build test:

```
led get-build 8820607344001876577 | led edit -d id=build831-m9 | led edit -d pool=luci.flutter.staging | led edit -p '$flutter/osx_sdk={"sdk_version": "13a233", "runtime_versions": ["ios-13-0"] }' | led launch -modernize
```

https://chromium-swarm.appspot.com/task?id=596f2f3a6feff410failed with the runtime not being found

**Comment by CaseyHillers** [comment]
@jmagman my run took <2 minutes to install the runtime 
![image](https://user-images.githubusercontent.com/2148558/156829959-a578c88f-b344-4bf2-98d7-0ee698b1b8db.png)

Are the runtimes worth caching?

**Comment by godofredoc** [comment]
Yes, the cipd package needs to be cached and the installation step will only symlink to the cache.

**Comment by jmagman** [comment]
> failed with the runtime not being found

You need ios-15 for that test (the default).

**Comment by CaseyHillers** [comment]
> Yes, the cipd package needs to be cached and the installation step will only symlink to the cache.

If the issue is the swarming caches are getting malformed, what benefit are we getting from it? I'm not proposing clearing the cipd cache, which contains the ~11 GB runtime zip.

> You need ios-15 for that test (the default).

Thanks! I ran `led get-build 8820607344001876577 | led edit -d id=build831-m9 | led edit -d pool=luci.flutter.staging | led edit -p '$flutter/osx_sdk={"sdk_version": "13a233", "runtime_versions": ["ios-15-0"] }' | led launch -modernize`

I have no updates other than I don't think it's a cache issue (I think @keyonghan is right with the ios-15 runtime having a bad configuration). I had the following led tests, but all failed:

https://chromium-swarm.appspot.com/task?id=596f44b68aeb2f10 for ios-15
https://chromium-swarm.appspot.com/task?id=596f51720733ee10 for a rerun
https://chromium-swarm.appspot.com/task?id=596f600a23887310 for a rerun with clobber=true

I'm going to remove bots from prod that are having the failures

**Comment by CaseyHillers** [comment]
It looks like the immediate issue of prod being impacted is resolved (staging is still impacted). I don't know enough about iOS to keep this bug on my plate, so unassigning myself.

**Comment by godofredoc** [comment]
> If the issue is the swarming caches are getting malformed, what benefit are we getting from it? I'm not proposing clearing the cipd cache, which contains the ~11 GB runtime zip.

Not sure I totally understand your comment. The current problem seems to be caused by first running with runtimes support which deletes all the default runtimes and creates symlinks to the runtimes in cache and then later trying to use the default runtime which won't reinstall the default runtime.

I believe this can be fixed by always specifying a runtime to force creating symlinks even for the default runtime.

**Comment by jmagman** [comment]
> I have no updates other than I don't think it's a cache issue (I think @keyonghan is right with the ios-15 runtime having a bad configuration).

Do we know it's actually there?  On a machine reproducing this can I please get the contents of `/opt/s/w/ir/cache/osx_sdk/XCode.app/Contents/Developer/Platforms/iPhoneOS.platform/Library/Developer/CoreSimulator/Profiles/Runtimes`?

**Comment by jmagman** [comment]
More:
> 1. On a machine where this is happening, is `Contents/Developer/Platforms/iPhoneOS.platform/Library/Developer/CoreSimulator/Profiles/Runtimes` empty?
> 2. In the original build where the Xcode cache was populated, does it show that `ios_runtime` download in the `install_xcode` step?
> 3. Is the `mac_toolchain` version correct on these machines? The runtime feature was introduced not that long ago. See https://github.com/flutter/flutter/issues/94780 and https://flutter-review.googlesource.com/c/recipes/+/23980

**Comment by jmagman** [comment]
> The current problem seems to be caused by first running with runtimes support which deletes all the default runtimes and creates symlinks to the runtimes in cache and then later trying to use the default runtime which won't reinstall the default runtime.


So why did this start happening in prod? As far as I know none of these machines ran with runtime support.

**Comment by zanderso** [comment]
This is causing ~every other engine roll to fail, (https://github.com/flutter/flutter/pull/99578) so I'm bumping this back up to P1.

**Comment by godofredoc** [comment]
@keyonghan @jmagman is this fixed? I took a look and all the tests seem to be passing correctly now.

**Comment by jmagman** [comment]
I think it's fixed, thanks for splitting the cache up @godofredoc!
Closing.  @zanderso if you see this again please reopen.

**Comment by keyonghan** [comment]
This is happening again on [build809-m9](https://chromium-swarm.appspot.com/bot?id=build809-m9) and [build797-m9](https://chromium-swarm.appspot.com/bot?id=build797-m9)

Failed builds: 
build809-m9: https://ci.chromium.org/ui/p/flutter/builders/prod/Mac%20build_tests_3_4/3863/overview
build797-m9: https://ci.chromium.org/ui/p/flutter/builders/prod/Mac%20build_tests_3_4/3864/overview

<img width="1169" alt="Screen Shot 2022-03-16 at 11 55 26 AM" src="https://user-images.githubusercontent.com/54558023/158665935-0c006d63-94bd-4861-a933-4c5b7b4f2c34.png">

**Comment by keyonghan** [comment]
I found similar pattern as last time: flakes on both bots happened right after a failed release `beta` run. Sounds cache was polluted after a release failed run and then caused Master builds failed to find the `runtime`.

**Comment by godofredoc** [comment]
Can you please link the beta build that caused the issue?

**Comment by keyonghan** [comment]
build809: https://luci-milo.appspot.com/ui/p/flutter/builders/prod/Mac%20beta%20verify_binaries_codesigned/406/overview
build797: https://luci-milo.appspot.com/ui/p/flutter/builders/prod/Mac%20beta%20verify_binaries_codesigned/405/overview

Both failed at step `Verify binaries codesigned` on builder `Mac beta verify_binaries_codesigned`

**Comment by godofredoc** [comment]
Can we run a quick check to validate your theory?

It will require to pick a healthy bot and run the verify_binaries and then "Mac build_tests_3_4" the expectation is that it will fail.

**Comment by godofredoc** [comment]
Not sure how codesigned is impacting xcode, specially if they are using different caches.

**Comment by christopherfujino** [comment]
Just merged a recipe change to fix this, however, I believe the bots may need to be manually recovered: https://flutter-review.googlesource.com/c/recipes/+/28064

**Comment by zanderso** [comment]
The tree is closed on this, so I have increased to P0.

**Comment by godofredoc** [comment]
All the new failures are coming from these bots:

```
[build797-m9](https://chromium-swarm.appspot.com/bot?id=build797-m9)
[build798-m9](https://chromium-swarm.appspot.com/bot?id=build798-m9)
[build803-m9](https://chromium-swarm.appspot.com/bot?id=build803-m9)
[build805-m9](https://chromium-swarm.appspot.com/bot?id=build805-m9)
[build809-m9](https://chromium-swarm.appspot.com/bot?id=build809-m9)
```

**Comment by godofredoc** [comment]
Caches for those builders have been cleaned.
